src/block-ref.test.ts

```diff
@@ -1,59 +1,620 @@
+                },
+                async ({ engine, provider }) => {
+                  stubProviderRequests(provider, [
+                    buildStubForBlockNumberRequest('0x100'),
+                  ]);
+
+                  await engine.handle({
+                    jsonrpc: '2.0',
+                    id: 1,
+                    method,
+                    params: buildMockParamsWithBlockParamAt(
+                      blockParamIndex,
+                      blockParam,
+                    ),
+                  });
+
+                  expect(finalMiddleware).toHaveBeenCalledWith(
+                    expect.objectContaining({
+                      params: buildMockParamsWithBlockParamAt(
+                        blockParamIndex,
+                        blockParam,
+                      ),
+                    }),
+                    expect.anything(),
+                    expect.anything(),
+                    expect.anything(),
+                  );
+                },
+              );
+            });
+          },
+        );
+      });
+    });
+  });
+
+  describe('when the RPC method does not take a block parameter', () => {
+    it('does not make a direct request through the provider, but proceeds to the next middleware', async () => {
+      const finalMiddleware = buildFinalMiddlewareWithDefaultResponse();
+
+      await withTestSetup(
+        {
+          configureMiddleware: ({ provider, blockTracker }) => {
+            return {
+              middlewareUnderTest: createBlockRefMiddleware({
+                provider,
+                blockTracker,
+              }),
+              otherMiddleware: [finalMiddleware],
+            };
+          },
+        },
+        async ({ engine, provider }) => {
+          stubProviderRequests(provider, [
+            buildStubForBlockNumberRequest('0x100'),
+          ]);
+
+          await engine.handle({
+            jsonrpc: '2.0',
+            id: 1,
+            method: 'a_non_block_param_method',
+            params: ['some value', '0x200'],
+          });
+
+          expect(finalMiddleware).toHaveBeenCalledWith(
+            expect.objectContaining({
+              params: ['some value', '0x200'],
+            }),
+            expect.anything(),
+            expect.anything(),
+            expect.anything(),
+          );
+        },
+      );
+    });
+  });
+});
+
+/**
+ * Calls the given function, which should represent a test of some kind, with
+ * data that the test can use, namely, a JsonRpcEngine instance, a provider
+ * object, and a block tracker.
+ *
+ * @template T - The type that the function will return, minus the promise
+ * wrapper.
+ * @param options - Options.
+ * @param options.configureMiddleware - A function that is called to add the
+ * middleware-under-test, along with any other necessary middleware, to the
+ * engine.
+ * @param callback - A function.
+ * @returns Whatever the function returns.
+ */
+async function withTestSetup<T>(
```

Comment:
A lot of the helper functions here were copied from the tests for `retryOnEmpty`. I plan on introducing another PR after this one which will extract these helpers from the two files.